### ISSUE TITLE
OpenTelemetryCollector should have specific serviceAnnotations, podAnnotations, serviceLabels and podLabels settings

### ISSUE THREAD

**Issue by diranged** [open-telemetry/opentelemetry-operator#2884]
### Component(s)

collector

### Is your feature request related to a problem? Please describe.

Today the operator _copies_ the `annotations` and `labels` from the `OpenTelemetryCollector` objects to all of the resources it creates. This causes problems with tools like ArgoCD which use Annotations for resource tracking, so then the [`--annotations-filter`](https://github.com/open-telemetry/opentelemetry-operator/blob/main/main.go#L158C265-L158C298) flag was created.  This is an unideal hack though because it's an all or nothing thing... and it means that we propagate labels and annotations to lots of resources where they don't make sense. For example, the [`service.kubernetes.io/topology-mode`](https://kubernetes.io/docs/concepts/services-networking/topology-aware-routing/) label has to be added to all the resources (clutster roles, pods, deployments, etc) when we really only want it on the service.

### Describe the solution you'd like

I don't think that the Operator should automatically copy these Annotations or Labels at all ... and instead they should be defined as parameters.

**Comment by taer** [comment]
I see https://github.com/open-telemetry/opentelemetry-operator/blob/main/docs/api.md#opentelemetrycollectorspec-1 has an option for podAnnotations, but not podLabels

**Comment by Rohlik** [comment]
We would also love to see support for `podLabels` for our use case, where we set a `team` label based on the owner's team and use that later on to route alerts.

**Comment by shuheiktgw** [comment]
This totally makes sense to me. Let me see if I can help with this issue 🙂

**Comment by shuheiktgw** [comment]
I've created a PR to support pod labels. 
- https://github.com/open-telemetry/opentelemetry-operator/pull/3733

Once it's been merged, I'm going to check service[Annotations|Labels] as well 🙂

**Comment by shuheiktgw** [comment]
@iblancasa Let me move [our discussion](https://github.com/open-telemetry/opentelemetry-operator/pull/3733#issuecomment-2681200826) here! Thanks for the clarification. I feel that maintaining a list of annotations and labels that shouldn't be propagated to child resources within the operator may not be flexible enough, as some users might want to prevent the propagation of service- or environment-specific labels/annotations, while others may actually want to propagate those prohibited ones.

If you have any input, please let me know! @diranged @taer @Rohlik

**Comment by iblancasa** [comment]
> It seems that I also misunderstood this issue 🤦🏼. In my case, I would love to see the introduction of `podLabels` with the exact mechanism that already exists for `podAnnotations`, to be able to set custom labels to Collector and Target Allocator pods 🙏🏼.

That is fine. And maybe something we want to do too. But, from what I get as part of this issue, the idea is to not propagate some of the values.

I would propose adding the `podLabels` field as part of a separate issue.

**Comment by Rohlik** [comment]
You are right, so I just created https://github.com/open-telemetry/opentelemetry-operator/issues/3751.

**Comment by tonyhogsten** [comment]
This feature is quite important to us, as it would enable us to use workload identity authentication to send metrics to Azure Managed Prometheus. 

It requires the pod to have the label azure.workload.identity/use: "true" which seems impossible to set through the operator, so adding this would be very very helpful. Thanks 🙏

**Comment by blezoray** [comment]
I'm also interesting by this feature in order to use an Openshift feature that permits to generate TLS certificate for K8S service.
https://docs.redhat.com/en/documentation/openshift_container_platform/4.15/html/security_and_compliance/configuring-certificates#add-service-certificate_service-serving-certificate
To avoid any conflict, we must be able to configure each service individually (service, service-headless and service-monitoring)
```shell
$ oc get svc -l 'app.kubernetes.io/part-of=opentelemetry'
NAME                        TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)             AGE
test-collector              ClusterIP   172.30.125.69    <none>        4317/TCP,4318/TCP   22d
test-collector-headless     ClusterIP   None             <none>        4317/TCP,4318/TCP   22d
test-collector-monitoring   ClusterIP   172.30.161.157   <none>        8888/TCP            22d
```

**Comment by ChristianCiach** [comment]
When using ArgoCD with auto-sync and pruning enabled, ArgoCD will delete the generated `ClusterRole` and `ClusterRoleBinding` resources of the operator. The operator will then re-create the resources, and so on and so on.

This is only really a problem for cluster-wide resources, because these have no owner-references to the namespaced `OpenTelemetryCollector` objects. ArgoCD thankfully doesn't regard resources with OwnerReferences as managed by ArgoCD, even if they have the ArgoCD tracking labels/annotations. So this is usually no big deal for most resources, but ClusterRole(Binding)s are a huge issue.

Edit:

What even are the reasons for propagating the labels/annotations at all? Operators usually only propagate only a limited subset of annotations/labels that they actually care about, like the [DataDog agent that only propagates some (as they call it) "default labels".](https://github.com/DataDog/datadog-operator/blob/3f74c10507269279fc756327ed9cad8cfca1bcfc/internal/controller/datadogagent/store/store.go#L104)

Looking at the history, it looks like the original intent was allowing to override the default prometheus-annotations on the pod by adding annotations to the `OpenTelemetryCollector` object. This seems awkward and doesn't make a lot of sense today where the annotations/labels are not only propagated to the pods, but to all other generated objects as well.

**Comment by gustavovnicius** [comment]
Adding a concrete use case, since this thread has a lot of "I need this for X" but not much on what the absence actually costs.

**Use case:** annotating the operator-generated Services with `networking.istio.io/exportTo: "."` so they stay namespace-local in an Istio mesh.

**Why it matters for `mode: daemonset`.** For a DaemonSet collector the operator generates four Services, with different traffic policies:

| Service | Port | internalTrafficPolicy | Endpoints published to every sidecar |
|---|---|---|---|
| `<name>-collector` | 4317/4318 | `Local` | 1 |
| `<name>-collector-headless` | 4317/4318 | `Local` | none (headless, no EDS) |
| `<name>-collector-extension` | 13133 | `Cluster` | **one per node** |
| `<name>-collector-monitoring` | 8888 | `Cluster` | **one per node** |

Istio publishes the full service registry to every sidecar unless you scope it, so the two `Cluster`-policy Services get pushed to every proxy in the mesh at one endpoint per node each.

The asymmetry is the interesting part. The Service that applications actually send OTLP to costs a single endpoint per sidecar, because `internalTrafficPolicy: Local` is set and Istio honours it. The two that scale linearly with cluster size are `-extension` and `-monitoring`, and neither is ever called through the mesh: `:13133` is the health_check extension hit by the kubelet, `:8888` is the metrics endpoint scraped by Prometheus. Both scrapers sit outside the mesh.

So on any reasonably sized cluster the overwhelming majority of the collector's footprint in every sidecar is for two Services that no meshed workload will ever talk to. There is a churn cost on top of the static one: replacing a node changes those endpoints and triggers an EDS push to every proxy, which is most noticeable during cluster upgrades when nodes roll through.

**Why `podAnnotations` doesn't help.** It exists and works, but `exportTo` is a *Service*-level annotation, and no field on the CRD reaches the generated Services:

```console
$ kubectl explain otelcol.spec | grep -iE 'annotation|service'
  podAnnotations   <map[string]string>
  serviceAccount   <string>
  serviceName      <string>
```

Copying top-level metadata annotations onto every generated resource (the behaviour this issue is about) would not be a good workaround either, since `exportTo` on a Deployment, ServiceAccount or ClusterRole is meaningless at best.

**What would solve it.** Anything that lets an annotation reach only the generated Services. `spec.serviceAnnotations` as proposed would do it. Being able to target the auxiliary Services individually would be better still, since in this case those are the only ones needing scoping and the primary Service is already well-behaved.

Happy to help test a PR against a DaemonSet-mode deployment in a mesh if that's useful.
